include/openssl/hpke.h

```diff
@@ -0,0 +1,466 @@
+ * @param recippriv is the EVP_PKEY form of recipient private value
+ * @param info is the info parameter
+ * @param infolen is the size the above
+ * @param aad is the aad parameter
+ * @param aadlen is the size the above
+ * @param ct is the ciphertext output
+ * @param ctlen is the size the above
+ * @return 1 for success, 0 for error
+ *
+ * This can be called once, or multiple, times.
+ *
+ * The recipient private key is explicitly set here as recipients
+ * are likely to be servers with multiple long(ish) term private
+ * keys in memory at once and that may have to e.g. do trial
+ * decryptions.
+ *
+ * The plaintext output (pt) will be smaller than the
+ * ciphertext input for all supported suites.
+ */
+int OSSL_HPKE_recipient_open(OSSL_HPKE_CTX *ctx,
+                             unsigned char *pt, size_t *ptlen,
+                             const unsigned char *enc, size_t enclen,
+                             EVP_PKEY *recippriv,
+                             const unsigned char *info, size_t infolen,
+                             const unsigned char *aad, size_t aadlen,
+                             const unsigned char *ct, size_t ctlen);
+
+/**
+ * @brief generate a given-length secret based on context and label
+ * @param ctx is the HPKE context
+ * @param secret is the resulting secret that will be of length...
+ * @param secret_len is the desired output length
+ * @param label is a buffer to provide separation between secrets
+ * @param labellen is the length of the above
+ * @return 1 for good, 0 for error
+ *
+ * The context has to have been used already for one encryption
+ * or decryption for this to work (as this is based on the negotiated
+ * "exporter_secret" estabilshed via the HPKE operation).
+ */
+int OSSL_HPKE_CTX_export(OSSL_HPKE_CTX *ctx,
+                         unsigned char *secret,
+                         size_t secret_len,
+                         const unsigned char *label,
+                         size_t labellen);
+
+/**
+ * @brief generate a key pair
+ * @param libctx is the context to use (normally NULL)
+ * @param propq is a properties string
+ * @param mode is the mode (currently unused)
+ * @param suite is the ciphersuite (currently unused)
+ * @param ikm is IKM, if supplied
+ * @param ikmlen is the length of IKM, if supplied
+ * @param pub is the public value
+ * @param publen is the size of the public key buffer (exact length on output)
+ * @param priv is the private key
+ * @return 1 for success, other for error (error returns can be non-zero)
+ *
+ * Used for entities that will later receive HPKE values to
+ * decrypt or that want a private key for an AUTH mode. Currently,
+ * only the KEM from the suite is significant here.
+ * The ``pub`` output will typically be published so that
+ * others can encrypt to the private key holder using HPKE.
+ * (Or authenticate HPKE values from that sender.)
+ */
+int OSSL_HPKE_keygen(OSSL_LIB_CTX *libctx, const char *propq,
```

Comment:
I guess we need to figure out what passing a NULL ikm here is supposed to do.. Is it supposed to do a normal keygen or should it internally use the dhkem derivekey.

Reply:
with no ikm supplied we should randomly generate - I just end up calling EVP_PKEY_generate in that case